diff --git a/tests/conftest.py b/tests/conftest.py
index d1924cb90..78e0ee2c7 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -425,12 +425,13 @@ def get_db_session_for_app_config(app_config):
         Session.remove()
         trans.rollback()
         conn.close()
-        engine.dispose()
 
 
 @pytest.fixture(scope="session")
 def app_config(database):
-    return get_app_config(database)
+    cfg = get_app_config(database)
+    yield cfg
+    cfg.registry["sqlalchemy.engine"].dispose()
 
 
 @pytest.fixture(scope="session")
@@ -444,7 +445,9 @@ def app_config_dbsession_from_env(database):
         "sessions.url": "redis://redis:0/",
     }
 
-    return get_app_config(database, nondefaults)
+    cfg = get_app_config(database, nondefaults)
+    yield cfg
+    cfg.registry["sqlalchemy.engine"].dispose()
 
 
 @pytest.fixture
@@ -529,16 +532,25 @@ def organization_service(db_session):
 
 @pytest.fixture
 def billing_service(app_config):
+    original_api_base = stripe.api_base
+    original_api_version = stripe.api_version
+    original_api_key = stripe.api_key
+
     stripe.api_base = app_config.registry.settings["billing.api_base"]
     stripe.api_version = app_config.registry.settings["billing.api_version"]
     stripe.api_key = "sk_test_123"
-    return subscription_services.MockStripeBillingService(
+
+    yield subscription_services.MockStripeBillingService(
         api=stripe,
         publishable_key="pk_test_123",
         webhook_secret="whsec_123",
         domain="localhost",
     )
 
+    stripe.api_base = original_api_base
+    stripe.api_version = original_api_version
+    stripe.api_key = original_api_key
+
 
 @pytest.fixture
 def subscription_service(db_session):
@@ -678,9 +690,10 @@ def _enable_all_oidc_providers(webtest):
 @pytest.fixture
 def _enable_organizations(db_request):
     flag = db_request.db.get(AdminFlag, AdminFlagValue.DISABLE_ORGANIZATIONS.value)
+    original_value = flag.enabled
     flag.enabled = False
     yield
-    flag.enabled = True
+    flag.enabled = original_value
 
 
 @pytest.fixture
@@ -753,8 +766,14 @@ def tm():
     tm.abort()
 
 
+@pytest.fixture(scope="session")
+def _wsgi_app(app_config_dbsession_from_env):
+    app_config_dbsession_from_env.add_settings(enforce_https=False)
+    return app_config_dbsession_from_env.make_wsgi_app()
+
+
 @pytest.fixture
-def webtest(app_config_dbsession_from_env, tm):
+def webtest(_wsgi_app, app_config_dbsession_from_env, tm):
     """
     This fixture yields a test app with an alternative Pyramid configuration,
     injecting the database session and transaction manager into the app.
@@ -764,18 +783,13 @@ def webtest(app_config_dbsession_from_env, tm):
     After the fixture has yielded the app, the transaction is rolled back and
     the database is left in its previous state.
     """
-
-    # We want to disable anything that relies on TLS here.
-    app_config_dbsession_from_env.add_settings(enforce_https=False)
-
-    app = app_config_dbsession_from_env.make_wsgi_app()
     engine = app_config_dbsession_from_env.registry["sqlalchemy.engine"]
 
     with get_db_session_for_app_config(app_config_dbsession_from_env) as _db_session:
         # Register the app with the external test environment, telling
         # request.db to use this db_session and use the Transaction manager.
         testapp = _TestApp(
-            app,
+            _wsgi_app,
             engine,
             extra_environ={
                 "warehouse.db_session": _db_session,
diff --git a/warehouse/admin/services.py b/warehouse/admin/services.py
index d1a2fffaa..bf1ab6499 100644
--- a/warehouse/admin/services.py
+++ b/warehouse/admin/services.py
@@ -9,11 +9,7 @@ import google.api_core.retry
 from zope.interface import implementer
 
 from warehouse.admin.interfaces import ISponsorLogoStorage
-from warehouse.utils.exceptions import DevelopmentModeWarning
-
-
-class InsecureStorageWarning(DevelopmentModeWarning):
-    pass
+from warehouse.utils.exceptions import InsecureStorageWarning
 
 
 @implementer(ISponsorLogoStorage)
diff --git a/warehouse/packaging/services.py b/warehouse/packaging/services.py
index e98dc4c8c..93ee7ff6d 100644
--- a/warehouse/packaging/services.py
+++ b/warehouse/packaging/services.py
@@ -53,7 +53,7 @@ from warehouse.packaging.models import (
 )
 from warehouse.packaging.typosnyper import typo_check_name
 from warehouse.rate_limiting import DummyRateLimiter, IRateLimiter
-from warehouse.utils.exceptions import DevelopmentModeWarning
+from warehouse.utils.exceptions import DevelopmentModeWarning, InsecureStorageWarning
 from warehouse.utils.project import PROJECT_NAME_RE
 
 logger = logging.getLogger(__name__)
@@ -75,10 +75,6 @@ STDLIB_PROHIBITED = {
 }
 
 
-class InsecureStorageWarning(DevelopmentModeWarning):
-    pass
-
-
 class GenericLocalBlobStorage:
     def __init__(self, base):
         # This class should not be used in production, it's trivial for it to
diff --git a/warehouse/utils/exceptions.py b/warehouse/utils/exceptions.py
index fd1a1fba0..3457d50ea 100644
--- a/warehouse/utils/exceptions.py
+++ b/warehouse/utils/exceptions.py
@@ -5,6 +5,10 @@ class DevelopmentModeWarning(UserWarning):
     pass
 
 
+class InsecureStorageWarning(DevelopmentModeWarning):
+    pass
+
+
 class InsecureOIDCPublisherWarning(DevelopmentModeWarning):
     pass
 
