diff --git a/pyproject.toml b/pyproject.toml
index a6b46907a..a861d7401 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -190,19 +190,14 @@ ignore = [
 # N805: zope.interface definitions look like classes, but are not
 "interfaces.py" = ["N805"]
 
-# INP001: Our migrations are scripts not files that are part of a package
-"warehouse/migrations/**/*.py" = ["INP001"]
-
 # The warehouse cli can use print to give feedback to the user.
 # The warehouse cli also defers imports on purpose.
 "warehouse/cli/**/*.py" = ["T201", "PLC0415"]
 
-# INP001: These are just scripts as well.
-# T201: These scripts/helpers are fine to call print
-"bin/*.py" = ["INP001", "T201"]
-"dev/*.py" = ["INP001", "T201"]
-"docs/**/*.py" = ["INP001"]
-".github/codeql/custom-queries/python/**/*.py" = ["INP001"]
+# INP001: These are scripts, not files that are part of a package.
+# T201: These scripts/helpers are fine to call print.
+"{bin,dev}/*.py" = ["INP001", "T201"]
+"{docs,.github/codeql/custom-queries/python,warehouse/migrations}/**/*.py" = ["INP001"]
 
 # PT009: These files use a lot of custom assertion tools that trip up this detection.
 "tests/unit/email/test_init.py" = ["PT009"]
